packages/@lwc/engine-core/src/framework/vm.ts

```diff
@@ -249,6 +249,32 @@ export function removeVM(vm: VM) {
     resetComponentStateWhenRemoved(vm);
 }
 
+function getNearestShadowAncestor(vm: VM): VM | null {
+    let ancestor = vm.owner;
+    while (!isNull(ancestor) && ancestor.renderMode === RenderMode.Light) {
+        ancestor = ancestor.owner;
+    }
+    return ancestor;
+}
+
+function assertNotSyntheticComposedWithinNative(vm: VM) {
+    const isSynthetic =
+        vm.renderMode === RenderMode.Shadow && vm.shadowMode === ShadowMode.Synthetic;
+    if (!isSynthetic) {
+        return;
+    }
+    const ancestor = getNearestShadowAncestor(vm);
+    if (!isNull(ancestor)) {
+        // Any native shadow component being an ancestor of a synthetic shadow component is disallowed.
+        assert.isFalse(
+            ancestor.renderMode === RenderMode.Shadow && ancestor.shadowMode === ShadowMode.Native,
+            `<${vm.tagName}> (synthetic shadow DOM) cannot be composed inside of <${
+                getAssociatedVM(ancestor.elm).tagName
+            }> (native shadow DOM), because synthetic-within-native composition is disallowed`
```

Comment:
There is a dedicated helper function to format [component tag names](https://github.com/salesforce/lwc/blob/d93e2963d8d32b138ce9ed130d98e7bef84b3d74/packages/%40lwc/engine-core/src/shared/format.ts#L11-L13).

```suggestion
            ancestor.renderMode === RenderMode.Shadow && ancestor.shadowMode === ShadowMode.Native,
            `${getComponentTag(vm)} (synthetic shadow DOM) cannot be composed inside of ${
                getComponentTag(ancestor)
            } (native shadow DOM), because synthetic-within-native composition is disallowed`
```